// ngx-restangular/esm2015/lib/ngx-restangular-helper.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ngx-restangular-helper.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HttpRequest, HttpHeaders, HttpParams } from '@angular/common/http';
import { assign } from 'core-js/features/object';
export class RestangularHelper {
    /**
     * @param {?} options
     * @return {?}
     */
    static createRequest(options) {
        /** @type {?} */
        const requestQueryParams = RestangularHelper.createRequestQueryParams(options.params);
        /** @type {?} */
        const requestHeaders = RestangularHelper.createRequestHeaders(options.headers);
        /** @type {?} */
        const methodName = options.method.toUpperCase();
        /** @type {?} */
        const withCredentials = options.withCredentials || false;
        /** @type {?} */
        let request = new HttpRequest(methodName, options.url, options.data, {
            headers: requestHeaders,
            params: requestQueryParams,
            responseType: options.responseType,
            withCredentials
        });
        if (['GET', 'DELETE', 'HEAD', 'JSONP', 'OPTIONS'].indexOf(methodName) >= 0) {
            request = new HttpRequest(methodName, options.url, {
                headers: requestHeaders,
                params: requestQueryParams,
                responseType: options.responseType,
                withCredentials
            });
        }
        return request;
    }
    /**
     * @param {?} queryParams
     * @return {?}
     */
    static createRequestQueryParams(queryParams) {
        /** @type {?} */
        const requestQueryParams = assign({}, queryParams);
        /** @type {?} */
        let search = new HttpParams();
        for (const key in requestQueryParams) {
            /** @type {?} */
            let value = requestQueryParams[key];
            if (Array.isArray(value)) {
                value.forEach((/**
                 * @param {?} val
                 * @return {?}
                 */
                function (val) {
                    search = search.append(key, val);
                }));
            }
            else {
                if (typeof value === 'object') {
                    value = JSON.stringify(value);
                }
                search = search.append(key, value);
            }
        }
        return search;
    }
    /**
     * @param {?} headers
     * @return {?}
     */
    static createRequestHeaders(headers) {
        for (const key in headers) {
            /** @type {?} */
            const value = headers[key];
            if (typeof value === 'undefined') {
                delete headers[key];
            }
        }
        return new HttpHeaders(assign({}, headers));
    }
}
