// ngx-stripe/esm5/lib/interfaces/stripe-instance.interface.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/interfaces/stripe-instance.interface.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function StripeServiceInterface() { }
if (false) {
    /**
     * @return {?}
     */
    StripeServiceInterface.prototype.getInstance = function () { };
    /**
     * @param {?=} options
     * @return {?}
     */
    StripeServiceInterface.prototype.elements = function (options) { };
    /**
     * @param {?=} options
     * @return {?}
     */
    StripeServiceInterface.prototype.redirectToCheckout = function (options) { };
    /**
     * @param {?} clientSecret
     * @param {?=} data
     * @return {?}
     */
    StripeServiceInterface.prototype.confirmAuBecsDebitPayment = function (clientSecret, data) { };
    /**
     * @param {?} clientSecret
     * @param {?=} data
     * @param {?=} options
     * @return {?}
     */
    StripeServiceInterface.prototype.confirmBancontactPayment = function (clientSecret, data, options) { };
    /**
     * @param {?} clientSecret
     * @param {?=} data
     * @param {?=} options
     * @return {?}
     */
    StripeServiceInterface.prototype.confirmCardPayment = function (clientSecret, data, options) { };
    /**
     * @param {?} clientSecret
     * @param {?=} data
     * @param {?=} options
     * @return {?}
     */
    StripeServiceInterface.prototype.confirmEpsPayment = function (clientSecret, data, options) { };
    /**
     * @param {?} clientSecret
     * @param {?=} data
     * @param {?=} options
     * @return {?}
     */
    StripeServiceInterface.prototype.confirmFpxPayment = function (clientSecret, data, options) { };
    /**
     * @param {?} clientSecret
     * @param {?=} data
     * @param {?=} options
     * @return {?}
     */
    StripeServiceInterface.prototype.confirmGiropayPayment = function (clientSecret, data, options) { };
    /**
     * @param {?} clientSecret
     * @param {?=} data
     * @param {?=} options
     * @return {?}
     */
    StripeServiceInterface.prototype.confirmIdealPayment = function (clientSecret, data, options) { };
    /**
     * @param {?} clientSecret
     * @param {?=} data
     * @param {?=} options
     * @return {?}
     */
    StripeServiceInterface.prototype.confirmP24Payment = function (clientSecret, data, options) { };
    /**
     * @param {?} clientSecret
     * @param {?=} data
     * @return {?}
     */
    StripeServiceInterface.prototype.confirmSepaDebitPayment = function (clientSecret, data) { };
    /**
     * @param {?} clientSecret
     * @return {?}
     */
    StripeServiceInterface.prototype.handleCardAction = function (clientSecret) { };
    /**
     * @param {?} paymentMethodData
     * @return {?}
     */
    StripeServiceInterface.prototype.createPaymentMethod = function (paymentMethodData) { };
    /**
     * @param {?} clientSecret
     * @return {?}
     */
    StripeServiceInterface.prototype.retrievePaymentIntent = function (clientSecret) { };
    /**
     * @param {?} clientSecret
     * @param {?=} data
     * @return {?}
     */
    StripeServiceInterface.prototype.confirmAuBecsDebitSetup = function (clientSecret, data) { };
    /**
     * @param {?} clientSecret
     * @param {?=} data
     * @param {?=} options
     * @return {?}
     */
    StripeServiceInterface.prototype.confirmCardSetup = function (clientSecret, data, options) { };
    /**
     * @param {?} clientSecret
     * @param {?=} data
     * @return {?}
     */
    StripeServiceInterface.prototype.confirmSepaDebitSetup = function (clientSecret, data) { };
    /**
     * @param {?} clientSecret
     * @return {?}
     */
    StripeServiceInterface.prototype.retrieveSetupIntent = function (clientSecret) { };
    /**
     * @param {?} options
     * @return {?}
     */
    StripeServiceInterface.prototype.paymentRequest = function (options) { };
    /**
     * @param {?} tokenType
     * @param {?} data
     * @return {?}
     */
    StripeServiceInterface.prototype.createToken = function (tokenType, data) { };
    /**
     * @param {?} tokenType
     * @param {?=} data
     * @return {?}
     */
    StripeServiceInterface.prototype.createToken = function (tokenType, data) { };
    /**
     * @param {?} tokenType
     * @param {?} data
     * @return {?}
     */
    StripeServiceInterface.prototype.createToken = function (tokenType, data) { };
    /**
     * @param {?} tokenType
     * @param {?} data
     * @return {?}
     */
    StripeServiceInterface.prototype.createToken = function (tokenType, data) { };
    /**
     * @param {?} tokenType
     * @param {?=} element
     * @return {?}
     */
    StripeServiceInterface.prototype.createToken = function (tokenType, element) { };
    /**
     * @param {?} tokenType
     * @param {?} data
     * @return {?}
     */
    StripeServiceInterface.prototype.createToken = function (tokenType, data) { };
    /**
     * @param {?} tokenType
     * @param {?} data
     * @return {?}
     */
    StripeServiceInterface.prototype.createToken = function (tokenType, data) { };
    /**
     * @param {?} element
     * @param {?} sourceData
     * @return {?}
     */
    StripeServiceInterface.prototype.createSource = function (element, sourceData) { };
    /**
     * @param {?} sourceData
     * @return {?}
     */
    StripeServiceInterface.prototype.createSource = function (sourceData) { };
    /**
     * @param {?} source
     * @return {?}
     */
    StripeServiceInterface.prototype.retrieveSource = function (source) { };
    /**
     * @deprecated
     * @param {?} clientSecret
     * @param {?=} element
     * @param {?=} data
     * @return {?}
     */
    StripeServiceInterface.prototype.handleCardPayment = function (clientSecret, element, data) { };
    /**
     * @deprecated
     * @param {?} clientSecret
     * @param {?=} element
     * @param {?=} data
     * @return {?}
     */
    StripeServiceInterface.prototype.confirmPaymentIntent = function (clientSecret, element, data) { };
    /**
     * @deprecated
     * @param {?} clientSecret
     * @param {?=} element
     * @param {?=} data
     * @return {?}
     */
    StripeServiceInterface.prototype.handleCardSetup = function (clientSecret, element, data) { };
    /**
     * @deprecated
     * @param {?} clientSecret
     * @param {?=} element
     * @param {?=} data
     * @return {?}
     */
    StripeServiceInterface.prototype.confirmSetupIntent = function (clientSecret, element, data) { };
    /**
     * @deprecated
     * @param {?} clientSecret
     * @param {?=} element
     * @param {?=} data
     * @return {?}
     */
    StripeServiceInterface.prototype.handleFpxPayment = function (clientSecret, element, data) { };
}
