// ngx-stripe/esm5/lib/services/stripe.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/services/stripe.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Inject } from '@angular/core';
import { filter, map } from 'rxjs/operators';
import { STRIPE_PUBLISHABLE_KEY, STRIPE_OPTIONS } from '../interfaces/ngx-stripe.interface';
import { WindowRef } from './window-ref.service';
import { LazyStripeAPILoader } from './api-loader.service';
import { StripeInstance } from './stripe-instance.class';
var StripeService = /** @class */ (function () {
    function StripeService(key, options, loader, window) {
        this.key = key;
        this.options = options;
        this.loader = loader;
        this.window = window;
        if (key) {
            this.stripe = new StripeInstance(this.loader, this.window, key, options);
        }
    }
    /**
     * @return {?}
     */
    StripeService.prototype.getStripeReference = /**
     * @return {?}
     */
    function () {
        var _this = this;
        return this.loader.asStream().pipe(filter((/**
         * @param {?} status
         * @return {?}
         */
        function (status) { return status.loaded === true; })), map((/**
         * @return {?}
         */
        function () { return ((/** @type {?} */ (_this.window.getNativeWindow()))).Stripe; })));
    };
    /**
     * @return {?}
     */
    StripeService.prototype.getInstance = /**
     * @return {?}
     */
    function () {
        return this.stripe.getInstance();
    };
    /**
     * @param {?} key
     * @param {?=} options
     * @return {?}
     */
    StripeService.prototype.setKey = /**
     * @param {?} key
     * @param {?=} options
     * @return {?}
     */
    function (key, options) {
        return this.changeKey(key, options);
    };
    /**
     * @param {?} key
     * @param {?=} options
     * @return {?}
     */
    StripeService.prototype.changeKey = /**
     * @param {?} key
     * @param {?=} options
     * @return {?}
     */
    function (key, options) {
        this.stripe = new StripeInstance(this.loader, this.window, key, options);
        return this.stripe;
    };
    /**
     * @param {?=} options
     * @return {?}
     */
    StripeService.prototype.elements = /**
     * @param {?=} options
     * @return {?}
     */
    function (options) {
        return this.stripe.elements(options);
    };
    /**
     * @param {?=} options
     * @return {?}
     */
    StripeService.prototype.redirectToCheckout = /**
     * @param {?=} options
     * @return {?}
     */
    function (options) {
        return this.stripe.redirectToCheckout(options);
    };
    /**
     * @param {?} clientSecret
     * @param {?=} data
     * @return {?}
     */
    StripeService.prototype.confirmAuBecsDebitPayment = /**
     * @param {?} clientSecret
     * @param {?=} data
     * @return {?}
     */
    function (clientSecret, data) {
        return this.stripe.confirmAuBecsDebitPayment(clientSecret, data);
    };
    /**
     * @param {?} clientSecret
     * @param {?=} data
     * @param {?=} options
     * @return {?}
     */
    StripeService.prototype.confirmBancontactPayment = /**
     * @param {?} clientSecret
     * @param {?=} data
     * @param {?=} options
     * @return {?}
     */
    function (clientSecret, data, options) {
        return this.stripe.confirmBancontactPayment(clientSecret, data, options);
    };
    /**
     * @param {?} clientSecret
     * @param {?=} data
     * @param {?=} options
     * @return {?}
     */
    StripeService.prototype.confirmCardPayment = /**
     * @param {?} clientSecret
     * @param {?=} data
     * @param {?=} options
     * @return {?}
     */
    function (clientSecret, data, options) {
        return this.stripe.confirmCardPayment(clientSecret, data, options);
    };
    /**
     * @param {?} clientSecret
     * @param {?=} data
     * @param {?=} options
     * @return {?}
     */
    StripeService.prototype.confirmEpsPayment = /**
     * @param {?} clientSecret
     * @param {?=} data
     * @param {?=} options
     * @return {?}
     */
    function (clientSecret, data, options) {
        return this.stripe.confirmEpsPayment(clientSecret, data, options);
    };
    /**
     * @param {?} clientSecret
     * @param {?=} data
     * @param {?=} options
     * @return {?}
     */
    StripeService.prototype.confirmFpxPayment = /**
     * @param {?} clientSecret
     * @param {?=} data
     * @param {?=} options
     * @return {?}
     */
    function (clientSecret, data, options) {
        return this.stripe.confirmFpxPayment(clientSecret, data, options);
    };
    /**
     * @param {?} clientSecret
     * @param {?=} data
     * @param {?=} options
     * @return {?}
     */
    StripeService.prototype.confirmGiropayPayment = /**
     * @param {?} clientSecret
     * @param {?=} data
     * @param {?=} options
     * @return {?}
     */
    function (clientSecret, data, options) {
        return this.stripe.confirmGiropayPayment(clientSecret, data, options);
    };
    /**
     * @param {?} clientSecret
     * @param {?=} data
     * @param {?=} options
     * @return {?}
     */
    StripeService.prototype.confirmIdealPayment = /**
     * @param {?} clientSecret
     * @param {?=} data
     * @param {?=} options
     * @return {?}
     */
    function (clientSecret, data, options) {
        return this.stripe.confirmIdealPayment(clientSecret, data, options);
    };
    /**
     * @param {?} clientSecret
     * @param {?=} data
     * @param {?=} options
     * @return {?}
     */
    StripeService.prototype.confirmP24Payment = /**
     * @param {?} clientSecret
     * @param {?=} data
     * @param {?=} options
     * @return {?}
     */
    function (clientSecret, data, options) {
        return this.stripe.confirmP24Payment(clientSecret, data, options);
    };
    /**
     * @param {?} clientSecret
     * @param {?=} data
     * @return {?}
     */
    StripeService.prototype.confirmSepaDebitPayment = /**
     * @param {?} clientSecret
     * @param {?=} data
     * @return {?}
     */
    function (clientSecret, data) {
        return this.stripe.confirmSepaDebitPayment(clientSecret, data);
    };
    /**
     * @param {?} clientSecret
     * @return {?}
     */
    StripeService.prototype.handleCardAction = /**
     * @param {?} clientSecret
     * @return {?}
     */
    function (clientSecret) {
        return this.stripe.handleCardAction(clientSecret);
    };
    /**
     * @param {?} paymentMethodData
     * @return {?}
     */
    StripeService.prototype.createPaymentMethod = /**
     * @param {?} paymentMethodData
     * @return {?}
     */
    function (paymentMethodData) {
        return this.stripe.createPaymentMethod(paymentMethodData);
    };
    /**
     * @param {?} clientSecret
     * @return {?}
     */
    StripeService.prototype.retrievePaymentIntent = /**
     * @param {?} clientSecret
     * @return {?}
     */
    function (clientSecret) {
        return this.stripe.retrievePaymentIntent(clientSecret);
    };
    /**
     * @param {?} clientSecret
     * @param {?=} data
     * @return {?}
     */
    StripeService.prototype.confirmAuBecsDebitSetup = /**
     * @param {?} clientSecret
     * @param {?=} data
     * @return {?}
     */
    function (clientSecret, data) {
        return this.stripe.confirmAuBecsDebitSetup(clientSecret, data);
    };
    /**
     * @param {?} clientSecret
     * @param {?=} data
     * @param {?=} options
     * @return {?}
     */
    StripeService.prototype.confirmCardSetup = /**
     * @param {?} clientSecret
     * @param {?=} data
     * @param {?=} options
     * @return {?}
     */
    function (clientSecret, data, options) {
        return this.stripe.confirmCardSetup(clientSecret, data, options);
    };
    /**
     * @param {?} clientSecret
     * @param {?=} data
     * @return {?}
     */
    StripeService.prototype.confirmSepaDebitSetup = /**
     * @param {?} clientSecret
     * @param {?=} data
     * @return {?}
     */
    function (clientSecret, data) {
        return this.stripe.confirmSepaDebitSetup(clientSecret, data);
    };
    /**
     * @param {?} clientSecret
     * @return {?}
     */
    StripeService.prototype.retrieveSetupIntent = /**
     * @param {?} clientSecret
     * @return {?}
     */
    function (clientSecret) {
        return this.stripe.retrieveSetupIntent(clientSecret);
    };
    /**
     * @param {?} options
     * @return {?}
     */
    StripeService.prototype.paymentRequest = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        return this.stripe.paymentRequest(options);
    };
    /**
     * @param {?} tokenType
     * @param {?} data
     * @return {?}
     */
    StripeService.prototype.createToken = /**
     * @param {?} tokenType
     * @param {?} data
     * @return {?}
     */
    function (tokenType, data) {
        return this.stripe.createToken(tokenType, data);
    };
    /**
     * @param {?} a
     * @param {?=} b
     * @return {?}
     */
    StripeService.prototype.createSource = /**
     * @param {?} a
     * @param {?=} b
     * @return {?}
     */
    function (a, b) {
        return this.stripe.createSource(a, b);
    };
    /**
     * @param {?} source
     * @return {?}
     */
    StripeService.prototype.retrieveSource = /**
     * @param {?} source
     * @return {?}
     */
    function (source) {
        return this.stripe.retrieveSource(source);
    };
    /**
     * @deprecated
     */
    /**
     * @deprecated
     * @param {?} clientSecret
     * @param {?=} element
     * @param {?=} data
     * @return {?}
     */
    StripeService.prototype.handleCardPayment = /**
     * @deprecated
     * @param {?} clientSecret
     * @param {?=} element
     * @param {?=} data
     * @return {?}
     */
    function (clientSecret, element, data) {
        return this.stripe.handleCardPayment(clientSecret, element, data);
    };
    /**
     * @deprecated
     */
    /**
     * @deprecated
     * @param {?} clientSecret
     * @param {?=} element
     * @param {?=} data
     * @return {?}
     */
    StripeService.prototype.confirmPaymentIntent = /**
     * @deprecated
     * @param {?} clientSecret
     * @param {?=} element
     * @param {?=} data
     * @return {?}
     */
    function (clientSecret, element, data) {
        return this.stripe.confirmPaymentIntent(clientSecret, element, data);
    };
    /**
     * @deprecated
     */
    /**
     * @deprecated
     * @param {?} clientSecret
     * @param {?=} element
     * @param {?=} data
     * @return {?}
     */
    StripeService.prototype.handleCardSetup = /**
     * @deprecated
     * @param {?} clientSecret
     * @param {?=} element
     * @param {?=} data
     * @return {?}
     */
    function (clientSecret, element, data) {
        return this.stripe.handleCardSetup(clientSecret, element, data);
    };
    /**
     * @deprecated
     */
    /**
     * @deprecated
     * @param {?} clientSecret
     * @param {?=} element
     * @param {?=} data
     * @return {?}
     */
    StripeService.prototype.confirmSetupIntent = /**
     * @deprecated
     * @param {?} clientSecret
     * @param {?=} element
     * @param {?=} data
     * @return {?}
     */
    function (clientSecret, element, data) {
        return this.stripe.confirmSetupIntent(clientSecret, element, data);
    };
    /**
     * @deprecated
     */
    /**
     * @deprecated
     * @param {?} clientSecret
     * @param {?=} element
     * @param {?=} data
     * @return {?}
     */
    StripeService.prototype.handleFpxPayment = /**
     * @deprecated
     * @param {?} clientSecret
     * @param {?=} element
     * @param {?=} data
     * @return {?}
     */
    function (clientSecret, element, data) {
        return this.stripe.handleFpxPayment(clientSecret, element, data);
    };
    StripeService.decorators = [
        { type: Injectable }
    ];
    StripeService.ctorParameters = function () { return [
        { type: String, decorators: [{ type: Inject, args: [STRIPE_PUBLISHABLE_KEY,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [STRIPE_OPTIONS,] }] },
        { type: LazyStripeAPILoader },
        { type: WindowRef }
    ]; };
    return StripeService;
}());
export { StripeService };
if (false) {
    /** @type {?} */
    StripeService.prototype.stripe;
    /** @type {?} */
    StripeService.prototype.key;
    /** @type {?} */
    StripeService.prototype.options;
    /** @type {?} */
    StripeService.prototype.loader;
    /** @type {?} */
    StripeService.prototype.window;
}
