// ngx-stripe/esm5/lib/services/stripe-elements.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/services/stripe-elements.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { StripeService } from './stripe.service';
var StripeElementsService = /** @class */ (function () {
    function StripeElementsService(stripeService) {
        this.stripeService = stripeService;
    }
    /**
     * @param {?} stripe
     * @param {?=} options
     * @return {?}
     */
    StripeElementsService.prototype.elements = /**
     * @param {?} stripe
     * @param {?=} options
     * @return {?}
     */
    function (stripe, options) {
        if (options === void 0) { options = {}; }
        if (stripe) {
            if (Object.keys(options).length > 0) {
                return stripe.elements(options);
            }
            return stripe.elements();
        }
        else {
            if (Object.keys(options).length > 0) {
                return this.stripeService.elements(options);
            }
            return this.stripeService.elements();
        }
    };
    /**
     * @param {?} stripe
     * @param {?} options
     * @return {?}
     */
    StripeElementsService.prototype.paymentRequest = /**
     * @param {?} stripe
     * @param {?} options
     * @return {?}
     */
    function (stripe, options) {
        return stripe
            ? stripe.paymentRequest(options)
            : this.stripeService.paymentRequest(options);
    };
    /**
     * @param {?} options
     * @param {?} containerClass
     * @return {?}
     */
    StripeElementsService.prototype.mergeOptions = /**
     * @param {?} options
     * @param {?} containerClass
     * @return {?}
     */
    function (options, containerClass) {
        if (!containerClass || (options && options.classes)) {
            return options || {};
        }
        if (!options || !options.classes) {
            return tslib_1.__assign({}, (options || {}), { classes: {
                    base: containerClass,
                    complete: containerClass + "--complete",
                    empty: containerClass + "--empty",
                    focus: containerClass + "--focus",
                    invalid: containerClass + "--invalid",
                    webkitAutoFill: containerClass + "--webkit-autoFill"
                } });
        }
        return options || {};
    };
    StripeElementsService.decorators = [
        { type: Injectable }
    ];
    StripeElementsService.ctorParameters = function () { return [
        { type: StripeService }
    ]; };
    return StripeElementsService;
}());
export { StripeElementsService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    StripeElementsService.prototype.stripeService;
}
