// ngx-stripe/esm5/lib/components/payment-request-button.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/payment-request-button.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, ViewChild, ElementRef, EventEmitter, Output } from '@angular/core';
import { from } from 'rxjs';
import { StripeInstance } from '../services/stripe-instance.class';
import { StripeElementsService } from '../services/stripe-elements.service';
var StripePaymentRequestButtonComponent = /** @class */ (function () {
    function StripePaymentRequestButtonComponent(stripeElementsService) {
        this.stripeElementsService = stripeElementsService;
        this.load = new EventEmitter();
        this.change = new EventEmitter();
        this.blur = new EventEmitter();
        this.focus = new EventEmitter();
        this.ready = new EventEmitter();
        this.token = new EventEmitter();
        this.paymentMethod = new EventEmitter();
        this.source = new EventEmitter();
        this.cancel = new EventEmitter();
        this.shippingaddresschange = new EventEmitter();
        this.shippingoptionchange = new EventEmitter();
        this.notavailable = new EventEmitter();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    StripePaymentRequestButtonComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var options, elementsOptions, stripe, updateElements, elements;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        options = this.stripeElementsService.mergeOptions(this.options, this.containerClass);
                        elementsOptions = this.elementsOptions;
                        stripe = this.stripe;
                        updateElements = false;
                        if (!(changes.elementsOptions || changes.stripe || !this.elements)) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.stripeElementsService
                                .elements(stripe, elementsOptions)
                                .toPromise()];
                    case 1:
                        elements = _a.sent();
                        this.elements = elements;
                        updateElements = true;
                        _a.label = 2;
                    case 2:
                        if (changes.paymentOptions && this.paymentRequest) {
                            this.updateRequest(this.paymentOptions);
                        }
                        if (changes.options ||
                            changes.containerClass ||
                            !this.element ||
                            updateElements) {
                            if (this.element && !updateElements) {
                                this.update(options);
                            }
                            else if (this.elements && updateElements) {
                                this.paymentRequest = this.stripeElementsService.paymentRequest(stripe, this.paymentOptions);
                                this.paymentRequest.on('token', (/**
                                 * @param {?} ev
                                 * @return {?}
                                 */
                                function (ev) { return _this.token.emit(ev); }));
                                this.paymentRequest.on('paymentmethod', (/**
                                 * @param {?} ev
                                 * @return {?}
                                 */
                                function (ev) {
                                    return _this.paymentMethod.emit(ev);
                                }));
                                this.paymentRequest.on('source', (/**
                                 * @param {?} ev
                                 * @return {?}
                                 */
                                function (ev) { return _this.source.emit(ev); }));
                                this.paymentRequest.on('cancel', (/**
                                 * @return {?}
                                 */
                                function () { return _this.cancel.emit(); }));
                                this.paymentRequest.on('shippingaddresschange', (/**
                                 * @param {?} ev
                                 * @return {?}
                                 */
                                function (ev) {
                                    return _this.shippingaddresschange.emit(ev);
                                }));
                                this.paymentRequest.on('shippingoptionchange', (/**
                                 * @param {?} ev
                                 * @return {?}
                                 */
                                function (ev) {
                                    return _this.shippingoptionchange.emit(ev);
                                }));
                                this.element = this.elements.create('paymentRequestButton', tslib_1.__assign({ paymentRequest: this.paymentRequest }, options));
                                this.canMakePayment().subscribe((/**
                                 * @param {?} result
                                 * @return {?}
                                 */
                                function (result) {
                                    if (result) {
                                        _this.element.on('click', (/**
                                         * @param {?} ev
                                         * @return {?}
                                         */
                                        function (ev) { return _this.change.emit(ev); }));
                                        _this.element.on('blur', (/**
                                         * @return {?}
                                         */
                                        function () { return _this.blur.emit(); }));
                                        _this.element.on('focus', (/**
                                         * @return {?}
                                         */
                                        function () { return _this.focus.emit(); }));
                                        _this.element.on('ready', (/**
                                         * @return {?}
                                         */
                                        function () { return _this.ready.emit(); }));
                                        _this.element.mount(_this.stripeElementRef.nativeElement);
                                        _this.load.emit({
                                            paymentRequestButton: _this.element,
                                            paymentRequest: _this.paymentRequest
                                        });
                                    }
                                    else {
                                        _this.notavailable.emit();
                                    }
                                }));
                            }
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * @return {?}
     */
    StripePaymentRequestButtonComponent.prototype.canMakePayment = /**
     * @return {?}
     */
    function () {
        return from(this.paymentRequest.canMakePayment());
    };
    /**
     * @param {?} options
     * @return {?}
     */
    StripePaymentRequestButtonComponent.prototype.update = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        this.element.update(options);
    };
    /**
     * @param {?} options
     * @return {?}
     */
    StripePaymentRequestButtonComponent.prototype.updateRequest = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        this.paymentRequest.update(options);
    };
    /**
     * @return {?}
     */
    StripePaymentRequestButtonComponent.prototype.show = /**
     * @return {?}
     */
    function () {
        this.paymentRequest.show();
    };
    /**
     * @deprecated
     */
    /**
     * @deprecated
     * @return {?}
     */
    StripePaymentRequestButtonComponent.prototype.getButton = /**
     * @deprecated
     * @return {?}
     */
    function () {
        return this.element;
    };
    StripePaymentRequestButtonComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngx-stripe-payment-request-button',
                    template: "<div class=\"field\" #stripeElementRef></div>"
                }] }
    ];
    StripePaymentRequestButtonComponent.ctorParameters = function () { return [
        { type: StripeElementsService }
    ]; };
    StripePaymentRequestButtonComponent.propDecorators = {
        stripeElementRef: [{ type: ViewChild, args: ['stripeElementRef',] }],
        containerClass: [{ type: Input }],
        paymentOptions: [{ type: Input }],
        options: [{ type: Input }],
        elementsOptions: [{ type: Input }],
        stripe: [{ type: Input }],
        load: [{ type: Output }],
        change: [{ type: Output }],
        blur: [{ type: Output }],
        focus: [{ type: Output }],
        ready: [{ type: Output }],
        token: [{ type: Output }],
        paymentMethod: [{ type: Output }],
        source: [{ type: Output }],
        cancel: [{ type: Output }],
        shippingaddresschange: [{ type: Output }],
        shippingoptionchange: [{ type: Output }],
        notavailable: [{ type: Output }]
    };
    return StripePaymentRequestButtonComponent;
}());
export { StripePaymentRequestButtonComponent };
if (false) {
    /** @type {?} */
    StripePaymentRequestButtonComponent.prototype.stripeElementRef;
    /** @type {?} */
    StripePaymentRequestButtonComponent.prototype.element;
    /** @type {?} */
    StripePaymentRequestButtonComponent.prototype.paymentRequest;
    /** @type {?} */
    StripePaymentRequestButtonComponent.prototype.containerClass;
    /** @type {?} */
    StripePaymentRequestButtonComponent.prototype.paymentOptions;
    /** @type {?} */
    StripePaymentRequestButtonComponent.prototype.options;
    /** @type {?} */
    StripePaymentRequestButtonComponent.prototype.elementsOptions;
    /** @type {?} */
    StripePaymentRequestButtonComponent.prototype.stripe;
    /** @type {?} */
    StripePaymentRequestButtonComponent.prototype.load;
    /** @type {?} */
    StripePaymentRequestButtonComponent.prototype.change;
    /** @type {?} */
    StripePaymentRequestButtonComponent.prototype.blur;
    /** @type {?} */
    StripePaymentRequestButtonComponent.prototype.focus;
    /** @type {?} */
    StripePaymentRequestButtonComponent.prototype.ready;
    /** @type {?} */
    StripePaymentRequestButtonComponent.prototype.token;
    /** @type {?} */
    StripePaymentRequestButtonComponent.prototype.paymentMethod;
    /** @type {?} */
    StripePaymentRequestButtonComponent.prototype.source;
    /** @type {?} */
    StripePaymentRequestButtonComponent.prototype.cancel;
    /** @type {?} */
    StripePaymentRequestButtonComponent.prototype.shippingaddresschange;
    /** @type {?} */
    StripePaymentRequestButtonComponent.prototype.shippingoptionchange;
    /** @type {?} */
    StripePaymentRequestButtonComponent.prototype.notavailable;
    /** @type {?} */
    StripePaymentRequestButtonComponent.prototype.elements;
    /** @type {?} */
    StripePaymentRequestButtonComponent.prototype.stripeElementsService;
}
