// ngx-stripe/esm2015/lib/components/card-number.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/card-number.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, ViewChild, ElementRef, EventEmitter, Output, Optional } from '@angular/core';
import { StripeCardGroupDirective } from '../directives/card-group.directive';
import { StripeElementsService } from '../services/stripe-elements.service';
export class StripeCardNumberComponent {
    /**
     * @param {?} stripeElementsService
     * @param {?} cardGroup
     */
    constructor(stripeElementsService, cardGroup) {
        this.stripeElementsService = stripeElementsService;
        this.cardGroup = cardGroup;
        this.load = new EventEmitter();
        this.blur = new EventEmitter();
        this.change = new EventEmitter();
        this.focus = new EventEmitter();
        this.ready = new EventEmitter();
        this.escape = new EventEmitter();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (changes.options || changes.containerClass) {
                this.setupElement('options');
            }
        });
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.cardGroup) {
            this.cardGroupSubscription = this.cardGroup.elements.subscribe((/**
             * @param {?} elements
             * @return {?}
             */
            (elements) => {
                this.elements = elements;
                this.setupElement('elements');
            }));
        }
        else {
            throw new Error('StripeCardNumberComponent must have StripeCardGroupDirective parent');
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.element) {
            this.element.destroy();
        }
        if (this.cardGroupSubscription) {
            this.cardGroupSubscription.unsubscribe();
        }
    }
    /**
     * @param {?} options
     * @return {?}
     */
    update(options) {
        this.element.update(options);
    }
    /**
     * @deprecated
     * @return {?}
     */
    getCardNumber() {
        return this.element;
    }
    /**
     * @private
     * @param {?} source
     * @return {?}
     */
    setupElement(source) {
        /** @type {?} */
        const options = this.stripeElementsService.mergeOptions(this.options, this.containerClass);
        if (this.element && source === 'options') {
            this.update(options);
        }
        else if (this.elements && source === 'elements') {
            this.element = this.elements.create('cardNumber', options);
            this.element.on('change', (/**
             * @param {?} ev
             * @return {?}
             */
            (ev) => this.change.emit(ev)));
            this.element.on('blur', (/**
             * @return {?}
             */
            () => this.blur.emit()));
            this.element.on('focus', (/**
             * @return {?}
             */
            () => this.focus.emit()));
            this.element.on('ready', (/**
             * @return {?}
             */
            () => this.ready.emit()));
            this.element.on('escape', (/**
             * @return {?}
             */
            () => this.escape.emit()));
            this.element.mount(this.stripeElementRef.nativeElement);
            this.load.emit(this.element);
        }
    }
}
StripeCardNumberComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngx-stripe-card-number',
                template: `<div class="field" #stripeElementRef></div>`
            }] }
];
StripeCardNumberComponent.ctorParameters = () => [
    { type: StripeElementsService },
    { type: StripeCardGroupDirective, decorators: [{ type: Optional }] }
];
StripeCardNumberComponent.propDecorators = {
    stripeElementRef: [{ type: ViewChild, args: ['stripeElementRef',] }],
    containerClass: [{ type: Input }],
    options: [{ type: Input }],
    load: [{ type: Output }],
    blur: [{ type: Output }],
    change: [{ type: Output }],
    focus: [{ type: Output }],
    ready: [{ type: Output }],
    escape: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    StripeCardNumberComponent.prototype.stripeElementRef;
    /** @type {?} */
    StripeCardNumberComponent.prototype.element;
    /** @type {?} */
    StripeCardNumberComponent.prototype.containerClass;
    /** @type {?} */
    StripeCardNumberComponent.prototype.options;
    /** @type {?} */
    StripeCardNumberComponent.prototype.load;
    /** @type {?} */
    StripeCardNumberComponent.prototype.blur;
    /** @type {?} */
    StripeCardNumberComponent.prototype.change;
    /** @type {?} */
    StripeCardNumberComponent.prototype.focus;
    /** @type {?} */
    StripeCardNumberComponent.prototype.ready;
    /** @type {?} */
    StripeCardNumberComponent.prototype.escape;
    /** @type {?} */
    StripeCardNumberComponent.prototype.elements;
    /** @type {?} */
    StripeCardNumberComponent.prototype.cardGroupSubscription;
    /** @type {?} */
    StripeCardNumberComponent.prototype.stripeElementsService;
    /**
     * @type {?}
     * @private
     */
    StripeCardNumberComponent.prototype.cardGroup;
}
